docs/new-architecture-playbook.md

```diff
@@ -0,0 +1,2182 @@
+3. Migrating off `setNativeProps`
+4. Move the call to `requireNativeComponent` to a separate file
+5. Migrating off `dispatchViewManagerCommand`
+6. Using `codegenNativeComponent`
+
+#### Migrating `findNodeHandle` / getting a `HostComponent`
+
+<!-- alex ignore host -->
+
+Much of the migration work requires a HostComponent ref to access certain APIs that are only attached to host components (like View, Text, or ScrollView). HostComponents are the return value of calls to `requireNativeComponent`. `findNodeHandle` tunnels through multiple levels of component hierarchy to find the nearest native component.
+
+As a concrete example, this code uses `findNodeHandle` to tunnel from `MyComponent` through to the `View` rendered by `MyJSComponent`.
+
+```javascript
+class MyComponent extends React.Component<Props> {
+  _ref: ?React.ElementRef<typeof MyJSComponent>;
+
+  render() {
+    return <MyJSComponent ref={this._captureRef} onSubmit={this._onSubmit} />
+  }
+
+  _captureRef: (ref) => {
+    this._ref = ref;
+  }
+
+  _onSubmit: () => {
+    const nodeHandle = findNodeHandle(this._ref);
+
+    if (nodeHandle) {
+      UIManager.measure(nodeHandle, () => {});
+    }
+  }
+}
+
+function MyJSComponent(props) {
+  return (
+    <View>
+      <SubmitButton onSubmit={props.onSubmit} />
+    </View>
+  );
+}
+```
+
+We can’t convert this call to `this._ref.measure` because `this._ref` is an instance to `MyJSComponent`, which is not a HostComponent and thus does not have a `measure` function.
+
+`MyJSComponent` renders a View, which is a HostComponent, so we need to get a reference to `View` instead. There are typically two approaches to get what we need. If the component we need to get the ref from is a function component using `forwardRef` is probably the right choice. If it is a class component with other public methods, adding a public method for getting the ref is an option. Here are examples of those two forms:
+
+#### Using `forwardRef`
+
+```javascript
+class MyComponent extends React.Component<Props> {
+  _ref: ?React.ElementRef<typeof MyJSComponent>;
+
+  render() {
+    return <MyJSComponent ref={this._captureRef} onSubmit={this._onSubmit} />
+  }
+
+  _captureRef: (ref) => {
+    this._ref = ref;
+  }
+
+  _onSubmit: () => {
+    if (this._ref != null)
+      this._ref.measure(() => {});
+    }
+  }
+}
+
+const MyJSComponent = React.forwardRef((props, forwardedRef) => {
+  return (
+    <View ref={forwardedRef}>
+      <SubmitButton onSubmit={props.onSubmit} />
+    </View>
+  );
+});
+```
+
+#### Using a getter, (note the addition of `getViewRef`)
+
+```javascript
+class MyComponent extends React.Component<Props> {
+  _ref: ?React.ElementRef<typeof MyJSComponent>;
+
+  render() {
+    return <MyJSComponent ref={this._captureRef} onSubmit={this._onSubmit} />
+  }
+
+  _captureRef: (ref) => {
+    this._ref = ref;
+  }
+
+  _onSubmit: () => {
+    if (this._ref != null)
+      this._ref.**getViewRef****()**.measure(() => {});
```

Comment:
@cortinico is this intentional? :)

```
this._ref.getViewRef().measure(() => {});
```

Reply:
> @cortinico is this intentional? :)

Obviously now :/ Sadly, the tool we use internally is not working great exporting markdown and having to deal with codeblocks. I've already fixed some of those problems on the Android samples. Thanks for spotting this. Should be fixed now.